addon/components/base/bs-form.js

```diff
@@ -343,64 +343,67 @@ export default Component.extend({
+    let record;
 
+    this.incrementProperty('pendingSubmissions');
     this.get('onBefore')(model);
 
-    RSVP.resolve(this.get('hasValidator') ? this.validate(this.get('model')) : null)
-      .then(
-        (record) => {
-          return RSVP.resolve()
-            .then(() => {
-              if (this.get('hideValidationsOnSubmit') === true) {
-                this.set('showAllValidations', false);
-              }
-              return this.get('onSubmit')(model, record);
-            })
-            .then(
-              () => {
-                if (!this.get('isDestroyed')) {
-                  this.set('isSubmitted', true);
-                }
-              },
-              () => {
-                if (!this.get('isDestroyed')) {
-                  this.set('isRejected', true);
-                }
-              }
-            );
-        },
-        (error) => {
-          this.set('showAllValidations', true);
-
-          return RSVP.resolve()
-            .then(() => {
-              return this.get('onInvalid')(model, error);
-            })
-            .finally(() => {
-              if (!this.get('isDestroyed')) {
-                this.set('isRejected', true);
-              }
-            });
-        }
-      )
-      .finally(() => {
-        if (!this.get('isDestroyed')) {
-          if (this.get('pendingSubmissions') === 1) {
-            this.set('isSubmitting', false);
-          }
-
-          this.decrementProperty('pendingSubmissions');
-
-          // reset forced hiding of validations
-          if (this.get('showAllValidations') === false) {
-            this.set('showAllValidations', undefined);
-          }
-        }
-      });
+    try {
+      record = await (this.get('hasValidator') ? this.validate(model) : null);
+    } catch (error) {
+      // model is invalid
+      this.set('showAllValidations', true);
+
+      await this.get('onInvalid')(model, error);
+
+      if (this.get('isDestroyed')) {
+        return;
+      }
+
+      this.set('isRejected', true);
+      this.decrementProperty('pendingSubmissions');
+
+      throw error;
+    }
+
+    if (this.get('hideValidationsOnSubmit') === true) {
+      this.set('showAllValidations', false);
+    }
+
+    try {
+      await this.get('onSubmit')(model, record);
+
+      if (this.get('isDestroyed')) {
+        return;
+      }
+
+      this.set('isSubmitted', true);
+    } catch (error) {
+      if (this.get('isDestroyed')) {
+        return;
+      }
+
+      this.set('isRejected', true);
+
+      throw error;
+    } finally {
+      this.decrementProperty('pendingSubmissions');
+
+      // reset forced hiding of validations
+      if (this.get('showAllValidations') === false) {
+        this.set('showAllValidations', undefined);
+      }
+    }
+  },
+
+  async submit() {
+    try {
+      await this.submitHandler(...arguments);
+    } catch(err) {
+      // catch error throws by form's validation
+      // TODO: do not catch other errors
```

Comment:
Catching errors that are not validation errors might make debugging much more difficult. But I'm not sure how we should handle this one.

Reply:
What was the reason to catch here in the first place?

Reply:
We must not end up with uncaught errors in promises for submit event due to throwing validation errors. There are to options to achieve that one (if I didn't missed something):

- Not rethrowing validation errors in `submitHandler()` if it was called by submit event (`submit()` hook).
- Catching validation errors rethrown in `submitHandler()` in `submit()` event hook.

The second one is currently implemented but has the issue of catch all errors thrown by sumit handler. That makes debugging way more difficult as the errors aren't reported to developer console. So maybe the first option would be better? I twould look like:

```js
Component.extend({
  submitHandler(e, rethrowValidationErrors = true) {
    // ...

    try {
      record = await (this.get('hasValidator') ? this.validate(model) : null);
    } catch (error) {
       // ...

       if (!rethrowValidationErrors) {
         return;
       }

       throw error;
    }

    // ...
  },

  async submit(event) {
    // should not reject on validation error as this would end up
    // as an uncaught rejection
    await this.submitHandler(event, false);
  },

  actions {
    submit() {
      // should reject on validation errors
      return this.submitHandler();
    }
  }
});

Reply:
> So maybe the first option would be better?

Agree, we shouldn't catch all errors!